Sources/PackageExtension/PackageExtension.swift

```diff
@@ -0,0 +1,339 @@
+//
+// The way in which SwiftPM communicates with the package extension is an im-
+// plementation detail, but the way it currently works is that the extension
+// is compiled (in a very similar way to the package manifest) and then run in
+// a sandbox. Currently it is passed the JSON encoded input structure as the
+// last command line argument; however, it this will likely change to instead
+// read it from stdin, since that avoids any command line length limitation.
+// Any generated commands and diagnostics are emitted on stdout after a zero
+// byte; this allows regular output, such as print statements for debugging,
+// to be emitted to SwiftPM verbatim. SwiftPM tries to interpret any stdout
+// contents after the last zero byte as a JSON encoded output struct in UTF-8
+// encoding; any failure to decode it is considered a protocol failure. The
+// exit code of the compiled extension determines success or failure (though
+// failure to decode the output is also considered a failure to run the ex-
+// tension).
+
+
+/// Provides information about the target being built, as well as contextual
+/// information such as the paths of the directories to which commands should
+/// be configured to write their outputs. This information should be used as
+/// part of generating the commands to be run during the build.
+public let targetBuildContext: TargetBuildContext = TargetBuildContextFactory()
+
+/// Constructs commands to run during the build, including full command lines.
+/// All paths should be based on the ones passed to the extension in the target
+/// build context.
+public let commandConstructor: CommandConstructor = CommandConstructor()
+
+/// Emits errors, warnings, and remarks to be shown as a result of running the
+/// extension. After emitting one or more errors, the extension should return a
+/// non-zero exit code.
+public let diagnosticsEmitter = DiagnosticsEmitter()
+
+
+/// Provides information about the target being built, as well as contextual
+/// information such as the paths of the directories to which commands should
+/// be configured to write their outputs. This information should be used as
+/// part of generating the commands to be run during the build.
+public final class TargetBuildContext: Decodable {
+
+    /// The name of the target being built.
+    public let targetName: String
+    
+    /// The module name of the target. This is usually derived from the name,
+    /// but is possibly be customizable in the package manifest in some future
+    /// SwiftPM version).
+    public let moduleName: String
+    
+    /// The path of the target directory.
+    public let targetDir: Path
+    
+    /// That path of the package that contains the target.
+    public let packageDir: Path
+    
+    /// Absolute paths of the source files in the target. This might include
+    /// derived source files generated by other extensions).
+    public let sourceFiles: [Path]
```

Comment:
We are likely to want to refine the notion of source vs resource vs other files before enabling this.  Currently SwiftPM makes this determination before extensions have a chance to run — it would clearly be more flexible to have the extensions be a part of this determination.  This will be refined in future PRs.